### PR TITLE
Calculate omen buy amount locally

### PR DESCRIPTION
**Pull request by evangriffiths**

### CODE REVIEW

File: prediction_market_agent_tooling/markets/omen/omen.py [hunk 1]
```diff
@@ -580,6 +580,23 @@ def get_user_url(cls, keys: APIKeys) -> str:
...
+            no_outcome_pool_size=outcome_token_pool[OMEN_FALSE_OUTCOME],
+            fee=self.fee,
+        )
+        return TokenAmount(amount=amount, currency=self.currency)
+
+    def _get_buy_token_amount_from_smart_contract(
+        self, bet_amount: BetAmount, direction: bool
+    ) -> TokenAmount:
+        """
+        Note: this is only valid if the market instance's token pool is
```
Comment (kongzii) [inline prediction_market_agent_tooling/markets/omen/omen.py:597]: Shouldn't this comment be on the other method?

File: prediction_market_agent_tooling/markets/omen/omen.py [hunk 2]
```diff
@@ -1208,3 +1225,36 @@ def withdraw_wxdai_to_xdai_to_keep_balance(
...
+    yes_outcome_pool_size: float,
+    no_outcome_pool_size: float,
+    fee: float,
+) -> float:
+    """
+    Calculates the amount of outcome tokens received for a given investment
+
+    Taken from https://github.com/gnosis/conditional-tokens-market-makers/blob/6814c0247c745680bb13298d4f0dd7f5b574d0db/contracts/FixedProductMarketMaker.sol#L264
+    """
+    outcome_index = float(buy_direction)  # Convert boolean to index (0 or 1)
```
Comment (kongzii) [inline prediction_market_agent_tooling/markets/omen/omen.py:1242]: It's sad but markets have `index 0 == yes/true`, `index 1 == no/false`.

Is this logic here correct?

File: tests/markets/omen/test_omen.py [hunk 3]
```diff
@@ -243,3 +243,28 @@ def test_get_new_p_yes() -> None:
...
+
+
+def test_get_buy_token_amount() -> None:
+    """
+    Test that the two methods of calculating buy amount are equivalent for a
+    'live' market (i.e. where the token pool matches that of the current smart
+    contract state)
+    """
+    market = OmenAgentMarket.get_binary_markets(
+        limit=1,
```
Comment (kongzii) [inline tests/markets/omen/test_omen.py:255]: Instead of 1, could this be at least `10` and also iterate over buy_direction True or False? Just feels like this calculation could have some edge cases

File: prediction_market_agent_tooling/markets/omen/omen.py (hunk 2 above)
Comment (evangriffiths) [reply prediction_market_agent_tooling/markets/omen/omen.py]: Oops, these variables were unused - removed

File: prediction_market_agent_tooling/markets/omen/omen.py (hunk 1 above)
Comment (evangriffiths) [reply prediction_market_agent_tooling/markets/omen/omen.py]: Yes, switched!

File: tests/markets/omen/test_omen.py (hunk 3 above)
Comment (evangriffiths) [reply tests/markets/omen/test_omen.py]: fair point. I tested with a larger number locally, but thought it was quite slow, but actually there's a larger fixed overhead and it's not much slower to make the extra smart contract calls (5.2s for 1 vs 7.3s for 10)

**Review by kongzii — APPROVED** [review]

File: prediction_market_agent_tooling/markets/omen/omen.py [hunk 4]
```diff
@@ -1208,3 +1225,33 @@ def withdraw_wxdai_to_xdai_to_keep_balance(
     logger.info(
         f"Withdrew {need_to_withdraw} wxDai to keep the balance above the minimum required balance {min_required_balance}."
     )
+
+
+def get_buy_token_amount(
+    investment_amount: float,
+    buy_direction: bool,
```
Comment (gabrielfior) [inline prediction_market_agent_tooling/markets/omen/omen.py:1232]: Tiny ask - can we rename this to `outcome_to_be_bought` or similar? Anything with the word `outcome` would suffice for me.
